src/resolve/resolved-dns-scope.c

```diff
@@ -1393,3 +1391,31 @@ int dns_scope_remove_dnssd_services(DnsScope *scope) {
 
         return 0;
 }
+
+bool dns_scope_has_route_only_domains(DnsScope *scope) {
+
+        DnsSearchDomain *domain, *first;
+        bool route_only = false;
+
+        /* Check if the scope has any route-only domains except of "~.", i. e. whether it should only be used for particular
+         * domains */
+
+        if (scope->protocol != DNS_PROTOCOL_DNS)
+                return false;
+
+        if (scope->link)
+                first = scope->link->search_domains;
+        else
+                first = scope->manager->search_domains;
+
+        LIST_FOREACH(domains, domain, first) {
+                /* "." means "any domain", thus the interface takes any kind of traffic. */
+                if (dns_name_is_root(DNS_SEARCH_DOMAIN_NAME(domain)))
+                        return false;
```

Comment:
Unless I'm misunderstanding the comment above, this does not match the comment. Should this be `continue` instead to match the comment?

The comments in users of this are:
`"Scope of DNS server %s has only route-only domains, not using as global name server"`
`/* Otherwise check if we have any route-only domains: if so, let's not volunteer as default route. */`
They seem to be incompatible :(

Reply:
So, this is not actually new code, but just some minor reordering of the old one, behaviour is the same.

I think the automatic logic of "let's not use this link as default route if it has any route-only domains set" was not necessarily the greatest logic ever, but it's also not totally flawed, and maybe was OK without this new explicit bool flag this patch set adds. So my intention was to retain the original logic as implicit default, already for compat reasons, and it works the way it works after all..

So this early exit and returning "false" even makes sense: after all the call is supposed to discern two kinds of links:

1. links that shall get only lookups for specific domains (i.e. your vpn link)
2. links that shall get any lookups (i.e. your internet link)

In the first case we return true, in the second case false.

Now, if you have "~." on as search domain on a link, then this should rightfully be considered of type 2. Hence the check does the right thing. After all if you have "~." on a link, then it doesn't really matter what other search domains it has too much, after all you explicitly said: "send me everything" that way...

Hope this makes sense?

I will extend the comment a bit more, about this.